### DISCUSSION TITLE
New [Experimental] attribute with language/compiler support

### DISCUSSION THREAD

**Discussion by jskeet** [dotnet/csharplang#3555] [Language Ideas] [answered]
We have a pretty decent way of indicating that a member in a library should no longer be used, with language/compiler support: `[Obsolete]`
However, we have no current way of indicating that a particular member is *experimental*, i.e. that it might be changed or removed in future versions.

Library authors currently have three choices:

- Create a new major version every time they want to break an experimental API signature: this hurts users who _don't_ use experimental API signatures
- Use documentation to indicate that something is experimental, and ignore breakage here when considering SemVer rules: this hurts users who don't always read all documentation thoroughly (so, everyone) and makes tooling around SemVer harder to automate
- Only include experimental members in non-GA releases: this complicates release processes significantly, particularly if there are reasonably long-running experimental features

Proposal: introduce a new attribute called `ExperimentalAttribute` - possibly with additional properties indicating the degree of experimentation, if the community thinks that's useful - and a compiler flag to allow the use of experimental features.

By default, trying to use an experimental API surface would generate an error, and IDEs should either not include them in autocomplete or only do so with an additional "but you need to enable..." indication. Specifying an "allow experimental feature use" would allow users to opt into using these members.

The generated code should not be affected by this at all, beyond the normal changes based on the presence of members. (So compiling the same code with and without the flag *could* generate different IL due to overload resolution for example, but no specific code generator changes would be needed.)

Nice-to-have options to be discussed:

- Allow it to be set temporarily via a pragma, effectively like ignoring a warning for a region of code
- Allow a value of "allow and check that I'm actually using it" - the intention behind this would be to allow consumers to opt into experimental features but then be informed when the feature is stable and they no longer need the flag.
- Allow particular experiment names, a bit like `ConditionalAttribute`, so that consumers could opt into *specific* experimental features without seeing others

**Comment by jskeet** [comment]
(Apologies if this is a duplicate - I did look beforehand, but didn't find anything.)

**Comment by YairHalberstadt** [comment]
Couldn't warnings be accomplished via analyzers, and hiding the members via `EditorBrowsable`?

**Comment by YairHalberstadt** [comment]
Source generators could be used to automatically generate the Editor browsable attribute from the experimental attribute by adding it to an enhanced partial method (c# 9 feature).

**Comment by ufcpp** [comment]
not in this repo but in the [runtime repo](https://github.com/dotnet/runtime/issues/31542)

**Comment by jskeet** [comment]
@YairHalberstadt: Given that we'd need a common attribute to start with (unless we decide everyone should create their own, which I'd really like to avoid) I think it makes more sense for the compiler to know about this than to rely on analyzers.

**Comment by jskeet** [comment]
@ufcpp: Wonderful, thank you - subscribed :)

**Comment by Turnerj** [comment]
> and a compiler flag to allow the use of experimental features

Would this somewhat act like how we can't have unsafe code without the csproj file having `AllowUnsafeBlocks`?

**Comment by YairHalberstadt** [comment]
@jskeet then why not add it to the standard library?

**Comment by YairHalberstadt** [comment]
I don't see why language/compiler support is necessary.

**Comment by jskeet** [comment]
@YairHalberstadt: Yes, the attribute *should* be added to the standard library. But I believe the feature would only really be a first class feature if it could be relied upon via the compiler rather than people using an analyzer.

**Comment by jskeet** [comment]
@Turnerj: Yes, "unsafe" is exactly the kind of model I'm thinking about.

**Comment by YairHalberstadt** [comment]
The analyzer and source generator could be included in the standard library as well.

I think it makes sense to build a precedent where features can be added in a first class way without compiler support, rather than requiring everything to be a new language feature.

**Comment by jskeet** [comment]
As I've just noted in the [runtime repo issue](https://github.com/dotnet/runtime/issues/31542) it may well be that versioning makes this infeasible. (You'd want to be able to remove experimental-in-1.1 features from 1.2, but then diamond dependencies would mean you might end up with 1.2 anyway, at which point it all breaks.

**Comment by jskeet** [comment]
@YairHalberstadt: What benefit do you see over "we don't need to write the language specification part"? Because I see that latter as actually a *benefit* - I'd far rather rely on something that is well-defined in the language specification than a piece of analyzer documentation which is entirely separate. I'd want to be able to rely on this in the same way that I rely on `[Obsolete]`.

**Comment by YairHalberstadt** [comment]
The advantage is that changing the compiler is hard and experimentation slow.

Doing this via already existing mechanisms means anyone can prototype this, it can be used via a nuget package, and then once it's stable it can be merged into dotnet/runtime.

**Comment by YairHalberstadt** [comment]
The biggest limitation is that any language changes have to go through the LDM, which is a huge bottleneck on new features, and has a nigh on infinite backlog, just like the compiler team.

If we can start adding some features through a completely seperate, community driven pipeline, I think doing so is almost a no brainer.

**Comment by jskeet** [comment]
I think something like this *ought* to go through the LDM. I'm proposing (or at least was proposing before the diamond dependency problem) something which I'd expect to have status-parity with `Obsolete`. I'd want the deliberation of the LDM. Yes, that slows things down - but I appreciate the uniformity of C# design precisely because it's thoughtful and not "let's just go for it".

*Prototyping* via an analyzer may well be a fantastic approach - but I'd want the final feature to be part of the language specification, with all the hoops that requires the feature to jump through.

**Comment by YairHalberstadt** [comment]
I think maybe a good approach, would be to create the prototype first, and then see if there's still something missing which necessitates language changes, or if the itch has been sufficiently scratched?

**Comment by jskeet** [comment]
@YairHalberstadt: Yes, that would be fine, but I suspect we may very well disagree about whether a language change would be a good idea at the end, if the prototype were successful. I *suspect* the diamond dependency issue is a fatal blow though :(

**Comment by 333fred** [comment]
Fascinatingly, this came up in offline discussion today on our teams chat @chsienki @RikkiGibson. Also, this kinda already exists, albeit in a wpf namespace: https://github.com/dotnet/roslyn/blob/fab7134296816fc80019c60b0f5bef7400cf23ea/docs/features/ExperimentalAttribute.md.

**Comment by 333fred** [comment]
It even works: https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATARgFgAobABgAIB1ASwDsUB7AdygDoAxRgVwYhhqM6bALJwYEFPwgBuEiToQwcKAAcIAYzjV6TVpx58BQ0eMnSSAbxIVbFANoBBGDAQ1g3eAFUoEAOZwABQ2dqHOru6ecAAqEAgBMOwAwgA20FAUmBThbh7wsfHi7ADKrtwaMJnZLrlRBQnsAKJ03GCZIaG2OZH5cQ1sAJJ08AgAZpraWd15MX1FbAAicClwfvxwSB2djikpLCLcKQKqKxQAvBTjKVBwAJQAulvYAMwUNxArKBR4FI0AHqpEDRlMMPtMohQQNUIjMttZiKEAL4kZHEEj2f6AtwgiQpR6kV4/JIUeGhF7ffAANm+ABYKCJArcSVtUaiSOSfgsrE9CXSGbc4VtQklRIy5Ai7KygA==

**Comment by anthonydotnet** [comment]
Food for thought... Instead of Experimental,  call this Unsafe ?

**Comment by jskeet** [comment]
@anthonydotnet: I'd be very strongly against that. Unsafe already has a very specific meaning in C# - let's not overload it.

**Comment by anthonydotnet** [comment]
> @anthonydotnet: I'd be very strongly against that. Unsafe already has a very specific meaning in C# - let's not overload it.

Hmm... yeah true.

**Comment by svick** [comment]
The analogy with unsafe is even stronger when you consider the proposed `[CallerMustBeUnsafe]` attribute (https://github.com/dotnet/roslyn-analyzers/issues/972). In both cases, you have:

* An attribute marking an API as dangerous in some way (`[Experimental]`, `[CallerMustBeUnsafe]`).
* A way of marking code that is allowed to use the dangerous API (compiler flags for both and also the `unsafe` keyword for unsafe).

I wonder if these two features could learn from each other and, if they were both adopted, if they should be made consistent, at least to some degree.

**Comment by sharwell** [comment]
Borrowed from an old [Preliminary Features](http://tunnelvisionlabs.github.io/dotnet-threading/docs-latest/html/e02b9881-f06b-4420-aace-71b176d635db.htm) section I wrote and modified to use the term 'experimental':

> This library may include types and/or members which are designated as experimental. The experimental designation is indicated by the application of `ExperimentalAttribute`. While experimental features are much more "flexible" during the release cycle of the library, certain rules do apply in order to ensure the stronger rules provided for stable features will not be violated by a change to an experimental feature. The following list includes examples of these rules, but other rules may be imposed by basic logical constraints. The API elements referred to in the following list are assumed to be restricted to the publicly-exposed API of the library.
>
> * A member may only refer to an experimental type in its signature if it is also marked experimental.
> * An interface may only include an experimental member if it is also marked experimental.
> * An interface may only extend an experimental interface if it is also marked experimental.
> * A class may only include an experimental `abstract` member if either it is also marked experimental, or all constructors for the class are marked `internal`, `private protected`, and/or `private`. This restriction also applies to abstract classes which do not implement an abstract member declared in a base class.

**Comment by JustNrik** [comment]
I would personally prefer an abstract WarningAttribute, where attributes that inherit from it will have a compiler warning, so we compiler team wouldn't have to create specific cases for any attribute like this in the compiler.

It would look like:

```cs
public abstract class WarningAttribute : Attribute
{
    public string Message { get; }

    public WarningAttribute(string message)
    {
        Message = message;
    }
}

[AttributeUsage(...)]
public sealed class ExperimentalAttribute : WarningAttribute
{
    public ExperimentalAttribute(string message) : base(message)
    {
    }
}

// Same for Obsolete/Deprecated/whatever. Compiler will just display Message property
// The compiler may as well enforce the attribute to have a constant expression for Message property I guess?
```

**Comment by HaloFour** [comment]
@JustNrik 

I like the idea, but I think the one problem with that approach is that currently `ObsoleteAttribute` generates a specific diagnostic code.  How would you differentiate the diagnostic codes from these different attributes?

**Comment by sharwell** [comment]
@HaloFour The diagnostic ID is configurable starting with https://github.com/dotnet/runtime/issues/33089

**Comment by Thaina** [comment]
Could this be covered in generalized code generation from packages?

Not sure which issue but I think if we could have ability to analyze and inject code, also ability to generate warning and/or error from package, `[Experimental]` attribute could also be another package or just use the same pattern in corefx

**Comment by terrajobst** [comment, marked as answer]
Would this proposal address this?

https://github.com/dotnet/designs/pull/285

**Reply by jskeet** [reply]
> Would this proposal address this?
> 
> [dotnet/designs#285](https://github.com/dotnet/designs/pull/285)

Quite possibly - thanks for highlighting it. I'll look more carefully later in the week, and no doubt have some feedback.

**Reply by jskeet** [reply]
Having finally reviewed this (sorry for the wait) I'd be happy for this discussion to be closed (assuming there *is* such a concept for discussions).

Thanks for the pointer!

**Reply by jodydonetti** [reply]
Damn, I just found out you finally made this thanks to a [video](https://youtu.be/Gv2uBJzBAms?si=8mM4_bbzvpwSZ8N2) from Nick Chapsas.

As the author of the [original proposal](https://github.com/dotnet/runtime/issues/31542) I would've liked to contribute to the discussion 🥲

Anyway, happy that it finally got included: I'll play with it as soon as it will be released officially.
